### PR TITLE
Challenge Code: Find All Unique Triplets That Sum to Zero

### PR DESCRIPTION
**Pull request by anwar099**
New Challenge:
Find All Unique Triplets That Sum to Zero
Write a function that finds all unique triplets in an array that sum up to zero. The function should return the triplets without duplicates.

---
name: solution review 
about: A template PR for code review with a checklist
---

# Behavior

## Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [ ] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [ ] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [ ] Each assumption about an argument is checked with an assertion
  - [ ] Each assertion checks for _only one_ assumption about the argument
- [ ] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
  - _It's ok to have extra helper functions if necessary, like with mergesort_
- [x] The function has correct type annotations
- [x] The function is not called at the top level of the function file 
  - _Recursive solutions **can** call the function from **inside** the function body_

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in any docstrings or tests
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

**Comment by Shahroz657** [comment]
Hey Anwar,

For py_formatting, please use ruff and black fromatter. And, please let me know if you need help with anything

**Comment by mudassra-taskeen** [comment]
Hey Anwar, You didn't add a unit test file of your problem. Please add that file too.

File: solutions/tests/test_three_sum.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        nums = [-2, 0, 0, 2, 2]
+        expected = [[-2, 0, 2]]
+        self.assertEqual(sorted(three_sum(nums)), sorted(expected))
+
+    def test_large_negative_and_positive_numbers(self):
+        """It should handle large positive and negative numbers."""
+        nums = [-(10**6), 0, 10**6, -1, 1]
+        expected = [[-(10**6), 0, 10**6], [-1, 0, 1]]
+        self.assertEqual(sorted(three_sum(nums)), sorted(expected))
+
```
Comment (mudassra-taskeen) [inline solutions/tests/test_three_sum.py:50]: Hey Anwar, Add a test case to handle lists with the same number repeated multiple times. 
for example:
nums = [1, 1, 1, 1].

File: solutions/tests/test_three_sum.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        nums = [0, 0, 0]
+        expected = [[0, 0, 0]]
+        self.assertEqual(three_sum(nums), expected)
+
+    def test_empty_list(self):
+        """It should return an empty list for an empty input list."""
+        nums = []
+        expected = []
+        self.assertEqual(three_sum(nums), expected)
+
```
Comment (mudassra-taskeen) [inline solutions/tests/test_three_sum.py:32]: I like the thing that you tried to add a variety of edge cases, including empty input, duplicate triplets, all zeroes, and large numbers, which is commendable.

File: solutions/tests/test_three_sum.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    def test_fewer_than_three_elements(self):
+        """It should return an empty list if fewer than three elements are provided."""
+        nums = [1, -1]
+        expected = []
+        self.assertEqual(three_sum(nums), expected)
+
```
Comment (mudassra-taskeen) [inline solutions/tests/test_three_sum.py:38]: Here are my suggestions, which you can consider for further improvement:
You can also add input validation and performance test.
For instance 

- If the function handles non-integer inputs gracefully, include tests for invalid inputs, e.g., nums = [1, "a", None]
- Add a test case to handle very large input arrays to ensure performance (e.g., nums = [1, -1, 0] * 10000)

File: solutions/three_sum.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
+def three_sum(nums):
+    """
+    Find all unique triplets in the array that sum up to zero.
+
+    Parameters:
+        nums (List[int]): The input list of integers.
+
+    Returns:
+        List[List[int]]: A list of unique triplets that sum to zero.
+
```
Comment (mudassra-taskeen) [inline solutions/three_sum.py]: Here you need to include a Raises section for exceptions.
For example:
TypeError: If nums is not a list of integers.

Comment (mudassra-taskeen) [inline solutions/three_sum.py]: Please add more details about the assumptions for the input list (e.g., duplicates are allowed, and the input doesn't need to be pre-sorted).

File: solutions/three_sum.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        >>> three_sum([-1, 0, 1, 2, -1, -4])
+        [[-1, -1, 2], [-1, 0, 1]]
+        >>> three_sum([0, 1, 1])
+        []
+        >>> three_sum([0, 0, 0])
+        [[0, 0, 0]]
+    """
+    nums.sort()  # Sort the array
+    result = []
+
```
Comment (mudassra-taskeen) [inline solutions/three_sum.py:13]: In defensive assertions, your function does not validate inputs explicitly.
Here is my suggestion:

- ensure nums is a list of integers:

- if not isinstance(nums, list) or not all(isinstance(x, int) for x in nums):

    raise TypeError("Input must be a list of integers.")

File: solutions/three_sum.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    """
+    Find all unique triplets in the array that sum up to zero.
+
+    Parameters:
+        nums (List[int]): The input list of integers.
+
+    Returns:
+        List[List[int]]: A list of unique triplets that sum to zero.
+
+    Examples:
```
Comment (mudassra-taskeen) [inline solutions/three_sum.py]: You’ve included doctests in the docstring. This is great.

Kindly add more examples for edge cases

>>> three_sum([1, 1, -2])
[[-2, 1, 1]]
>>> three_sum([1, 2, -3, 4, -1, -1])
[[-3, 1, 2], [-1, -1, 2]]

**Comment by anwar099** [comment]
Dear mudassra-taskeen I added the test file, BUT I don't know who removed it!

**Comment by mudassra-taskeen** [comment]
> Dear mudassra-taskeen I added the test file, BUT I don't know who removed it!

No, your file is not removed. It's there. Just look at my comments and make changes where required

File: solutions/three_sum.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    nums.sort()  # Sort the array
+    result = []
+
+    for i in range(len(nums) - 2):
+        # Skip duplicate elements
+        if i > 0 and nums[i] == nums[i - 1]:
+            continue
+
+        left, right = i + 1, len(nums) - 1
+        while left < right:
```
Comment (mudassra-taskeen) [inline solutions/three_sum.py:19]: The nested structure of while inside the for loop is fine but might make debugging harder. You can consider the two-pointer logic in a helper function for clarity.

**Comment by anwar099** [comment]
Thank you for the suggestions! I appreciate the feedback and will make the following updates based on your recommendations:

**Add Test Case for Repeated Numbers:**
I will add a test case to handle lists where the same number is repeated multiple times, such as nums = [1, 1, 1, 1].

**Input Validation:**
I'll ensure the function gracefully handles non-integer inputs by adding a validation mechanism and corresponding test cases, e.g., nums = [1, "a", None].

**Performance Testing:**
I'll include a test case with very large input arrays, such as nums = [1, -1, 0] * 10000, to ensure the function performs well under heavy input sizes.

File: solutions/three_sum.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        []
+        >>> three_sum([0, 0, 0])
+        [[0, 0, 0]]
+    """
+    nums.sort()  # Sort the array
+    result = []
+
+    for i in range(len(nums) - 2):
+        # Skip duplicate elements
+        if i > 0 and nums[i] == nums[i - 1]:
```
Comment (mudassra-taskeen) [inline solutions/three_sum.py:15]: To increase efficiency, you could add a quick break if the current number nums[i] is greater than 0, as no triplet can sum to zero in that case. Otherwise, implementation already avoids unnecessary comparisons.

**Comment by mudassra-taskeen** [comment]
> Thank you for the suggestions! I appreciate the feedback and will make the following updates based on your recommendations:
> 
> **Add Test Case for Repeated Numbers:** I will add a test case to handle lists where the same number is repeated multiple times, such as nums = [1, 1, 1, 1].
> 
> **Input Validation:** I'll ensure the function gracefully handles non-integer inputs by adding a validation mechanism and corresponding test cases, e.g., nums = [1, "a", None].
> 
> **Performance Testing:** I'll include a test case with very large input arrays, such as nums = [1, -1, 0] * 10000, to ensure the function performs well under heavy input sizes.

That's great.

**Review by mudassra-taskeen — APPROVED** [review]
Overall, your solution was good. There were some raises and test uses that I wanted you to add to ensure the efficiency of your code.

**Comment by mudassra-taskeen** [comment]
I have reviewed this code and added comments. I have been waiting for long so that Anwar can do some crucial modifications to his code. That's why I am late in merging this pull request since I was waiting to have some conversations on this problem and mark all checks.
